components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java

```diff
@@ -332,13 +337,84 @@ public AuthenticationContext authenticate(SignedJWTInfo signedJWTInfo, MessageCo
+                        if (isGatewayTokenCacheEnabled) {
+                            Object cachedTokenObj = getGatewayJWTTokenCache().get(jwtTokenCacheKey);
+                            if (cachedTokenObj instanceof String) {
+                                String cachedToken = (String) cachedTokenObj;
+                                long tsSkewMs = getTimeStampSkewInSeconds() * 1000;
+                                if (JWTUtil.isJWTValid(cachedToken, jwtConfigurationDto.getJwtDecoding(), tsSkewMs)) {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Using cached MCP upstream token for key: " + jwtTokenCacheKey);
+                                    }
+                                    internalToken = cachedToken;
+                                } else {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Cached MCP upstream token for key: " + jwtTokenCacheKey
+                                                + " expired, removing from cache");
+                                    }
+                                    getGatewayJWTTokenCache().remove(jwtTokenCacheKey);
+                                }
+                            }
+                        }
+                        if (StringUtils.isEmpty(internalToken)) {
+                            try {
+                                log.debug("Generating new MCP upstream token for existing API subtype");
+                                JwtTokenInfoDTO jwtTokenInfoDTO =
+                                        getJwtTokenInfoDTO(signedJWTInfo, jwtInfoDto, matchedAPI);
+                                internalToken = new InternalAPIKeyGenerator().generateToken(jwtTokenInfoDTO);
+                                if (isGatewayTokenCacheEnabled) {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("Caching generated MCP upstream token with key: "
+                                                + jwtTokenCacheKey);
+                                    }
+                                    getGatewayJWTTokenCache().put(jwtTokenCacheKey, internalToken);
+                                }
+                            } catch (APIManagementException e) {
+                                log.error("Error while generating MCP upstream token", e);
+                                throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+                                        APISecurityConstants.API_AUTH_GENERAL_ERROR_MESSAGE, e);
+                            }
+                        }
+                        authenticationContext.setMcpUpstreamToken(internalToken);
+                    } else if (matchedAPI == null) {
+                        log.warn("No matching MCP server found for token: " + GatewayUtils.getMaskedToken(jwtHeader));
```

Comment:
We need to rethink this error message. Token has noting to do with the matchedAPI not being returned right? We can fix this in a later PR as well if needed